pkg/storegateway/bucket_chunk_reader.go

```diff
@@ -242,6 +243,132 @@ type loadIdx struct {
 	chunk       int
 }
 
+type rangeLoadIdx struct {
+	offset uint32
+	estLen int64
+	// seriesEntry and rangeEntry are indices for where to put the result,
+	// they are not actual refs of series or chunks
+	seriesEntry int
+	rangeEntry  int
+}
+
+type bucketChunksRangesReader struct {
+	ctx    context.Context
+	block  *bucketBlock
+	toLoad [][]rangeLoadIdx
+}
+
+func newBucketChunksRangesReader(ctx context.Context, block *bucketBlock) *bucketChunksRangesReader {
+	return &bucketChunksRangesReader{
+		ctx:    ctx,
+		block:  block,
+		toLoad: make([][]rangeLoadIdx, len(block.chunkObjs)),
+	}
+}
+
+func (r *bucketChunksRangesReader) Close() error {
+	r.block.pendingReaders.Done()
+	return nil
+}
+
+// reset resets the ranges scheduled for loading. It does not release any loaded ranges.
+func (r *bucketChunksRangesReader) reset() {
+	for i := range r.toLoad {
+		r.toLoad[i] = r.toLoad[i][:0]
+	}
+}
+
+func (r *bucketChunksRangesReader) addLoadRange(g seriesChunkRefsRange, seriesEntry, rangeEntry int) error {
+	var (
+		seq = g.segmentFile
+		off = g.refs[0].segFileOffset
+	)
+	if seq >= uint32(len(r.toLoad)) {
+		return errors.Errorf("reference sequence %d out of range", seq)
+	}
+
+	var totalLen int64
+	for _, c := range g.refs {
+		totalLen += int64(c.length)
+	}
+
+	r.toLoad[seq] = append(r.toLoad[seq], rangeLoadIdx{offset: off, seriesEntry: seriesEntry, rangeEntry: rangeEntry, estLen: totalLen})
+	return nil
+}
+
+func (r *bucketChunksRangesReader) loadRanges(partialSeries []partialSeriesChunksSet, stats *safeQueryStats) error {
```

Comment:
[nit] May help to clarify that it's where we store the result.

```suggestion
func (r *bucketChunksRangesReader) loadRanges(result []partialSeriesChunksSet, stats *safeQueryStats) error {
```